src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/ProjectSystem/RazorProjectService.cs

```diff
@@ -19,28 +19,142 @@
-    : IRazorProjectService
+/// <summary>
+/// Maintains the language server's <see cref="IProjectSnapshotManager"/> with the semantics of Razor's project model.
+/// </summary>
+/// <remarks>
+/// This service implements both <see cref="IRazorStartupService"/> to ensure it is created early and
+/// <see cref="IOnInitialized"/> to ensure that its initialization completes when the language server
+/// finishes initialization.
+/// </remarks>
+internal partial class RazorProjectService : IRazorProjectService, IRazorProjectInfoListener, IRazorStartupService, IOnInitialized, IDisposable
 {
-    private readonly IProjectSnapshotManager _projectManager = projectManager;
-    private readonly RemoteTextLoaderFactory _remoteTextLoaderFactory = remoteTextLoaderFactory;
-    private readonly IDocumentVersionCache _documentVersionCache = documentVersionCache;
-    private readonly ILogger _logger = loggerFactory.GetOrCreateLogger<RazorProjectService>();
+    private readonly IRazorProjectInfoDriver _projectInfoDriver;
+    private readonly IProjectSnapshotManager _projectManager;
+    private readonly RemoteTextLoaderFactory _remoteTextLoaderFactory;
+    private readonly IDocumentVersionCache _documentVersionCache;
+    private readonly ILogger _logger;
+
+    private readonly CancellationTokenSource _disposeTokenSource;
+    private readonly Task _initializeTask;
+
+    public RazorProjectService(
+        IProjectSnapshotManager projectManager,
+        IRazorProjectInfoDriver projectInfoDriver,
+        IDocumentVersionCache documentVersionCache,
+        RemoteTextLoaderFactory remoteTextLoaderFactory,
+        ILoggerFactory loggerFactory)
+    {
+        _projectInfoDriver = projectInfoDriver;
+        _projectManager = projectManager;
+        _remoteTextLoaderFactory = remoteTextLoaderFactory;
+        _documentVersionCache = documentVersionCache;
+        _logger = loggerFactory.GetOrCreateLogger<RazorProjectService>();
+
+        // We kick off initialization immediately to ensure that the IRazorProjectService
+        // is hot and ready to go when requests come in.
+        _disposeTokenSource = new();
+        _initializeTask = InitializeAsync(_disposeTokenSource.Token);
+    }
 
-    public Task AddDocumentToMiscProjectAsync(string filePath, CancellationToken cancellationToken)
+    public void Dispose()
     {
-        return _projectManager.UpdateAsync(
-            updater: AddDocumentToMiscProjectCore,
-            state: filePath,
-            cancellationToken);
+        _disposeTokenSource.Cancel();
+        _disposeTokenSource.Dispose();
+    }
+
+    private async Task InitializeAsync(CancellationToken cancellationToken)
+    {
+        _logger.LogTrace($"Initializing {nameof(RazorProjectService)}...");
+
+        await _projectInfoDriver.WaitForInitializationAsync().ConfigureAwait(false);
+
+        // Register ourselves as a listener to the project driver.
+        _projectInfoDriver.AddListener(this);
```

Comment:
As per a previous comment, the initialization is odd here. This initialization happens immediately, it them initializes the project info driver, which goes through all of the current projects and starts a task for them, then it adds itself as a listener, then this goes through the current projects and updates them. Seems like both this and the driver are trying to inialize a starting set of projects, and if this `AddListener` call was moved above the `WaitForInitializationAsync` call then the below loop could be removed?

Reply:
In VS, the driver is "hot" and starts populating immediately. If AddListener is moved above WaitForInitializationAsync, I believe there's a potential race condition where incoming project updates could be processed earlier and then rolled back by the loop. For that reason, AbstractRazorProjectInfoDriver.AddListener() throws if called before initialization is finished.